### ISSUE TITLE
Nose-Hoover thermostat with SCF in Drude system

### ISSUE THREAD

**Issue by WangXinyan940** [openmm/openmm#3078]
While trying to do some non-equilibrim simulation, I found that DrudeNoseHoover thermostat sometimes shows funny results. Do I have any way to use Nose-Hoover thermostat together with SCF for Drude system? Can CompoundIntegrator help me do this?

**Comment by peastman** [comment]
Could you elaborate?  What sort of results are you talking about?  What exactly are you trying to do?

**Comment by WangXinyan940** [comment]
> Could you elaborate? What sort of results are you talking about? What exactly are you trying to do?

In each integration step, I want to couple Nose-Hoover thermostat with atoms and optimizing the positions of Drude particles using SCF. Just like what GROMACS do with shell particles.

**Comment by peastman** [comment]
I don't think there's any way to combine the two integrators like that.  But again, what sort of "funny" results were you talking about?

**Comment by jchodera** [comment]
Do you *need* to use a Nose-Hoover thermostat, or could you use Langevin instead?
If so, you could build a `CustomIntegrator` that performs Langevin dynamics on the nuclei and uses a simple gradient descent algorithm to optimize the Drude particles between each Langevin timestep.

**Comment by jchodera** [comment]
@WangXinyan940 : Just wondering if this was still a relevant issue for you, or if we can close it.

**Comment by vkwan8** [comment]
Bumping

Without writing a custom integrator, the only thermostat available for DrudeSCF is the Andersen thermostat (via AddForce). 

The Andersen thermostat, like the Langevin thermostat(*), does not reproduce correct dynamical properties. 

Maybe consider separating the Integrator class into Integrator and DrudeIntegrator?

**Comment by peastman** [comment]
Can you explain your use case?  If you're running constant temperature simulations, why do you want to use SCF rather than Langevin?

**Comment by vkwan8** [comment]
The main use case is to calculate dynamical properties (ie diffusion coefficient) for Drude models that does not support the use of the Langevin integrator. 

One example is the BK3 model, where one of the Drude particle reside on the water virtual particle. [Consequence: massless virtual particle means massless Drude particle means no extended Lagrangian methods].  

The traditional point-charged approach to Drude models does have some problem and I argue that some of the solutions are not that elegant. To solve the issue of polarization catastrophe, a counter charge is added to decrease the chances of running in to numerical instabilities which increases the number of charged interacting sites. For some models like SWM4-NDP, Thole Coulombic screen have to be used to prevent polarization catastrophe.

**Comment by peastman** [comment]
Ok, that makes sense.  I agree that doing these things with Drude particles involves workarounds that aren't always very elegant.

Be aware that DrudeSCFIntegrator is quite slow.  In principle it could be made faster with some effort.  We haven't bothered because Drude simulations usually use Langevin dynamics instead.

**Comment by dspoel** [comment]
Sorry to revive an old thread. Has anything happened with the SCF integrator? It would be great to be able to use it with a NH thermostat. It would remove the doubt about kinetics. I'm trying to implement our [force field for alkali halides](https://doi.org/10.1021/acs.jctc.8b00507) and while simulation can be made to run with the DrudeLangevinIntegrator we need to go to time steps of 0.4 fs, at 0.5 fs a sodium chloride crystal will melt (at room temperature). Using SCF in GROMACS 4.6.7 we were able to integrate the EOM with 2 fs, which should compensate the slowness of the integrator to a large extent.

![Image](https://github.com/user-attachments/assets/f3e7d256-0257-496d-b950-71b091b3f47f)

**Comment by peastman** [comment]
Yes, I implemented a faster algorithm in https://github.com/openmm/openmm/pull/4354.

**Comment by dspoel** [comment]
That's great, thanks. I will do some tests with it with the Andersen thermostat.
